pkg/controllers/sync/controller.go

```diff
@@ -1186,3 +1231,143 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
 
 	return worker.StatusAllOK
 }
+
+func (s *SyncController) listManagedClusterObjects(
+	ctx context.Context,
+	ftc *fedcorev1a1.FederatedTypeConfig,
+	clusterName string,
+) ([]*unstructured.Unstructured, sets.Set[string], error) {
+	clusterObjs := []*unstructured.Unstructured{}
+	namespaces := sets.New[string]()
+
+	resourceLister, hasSynced, exists := s.fedInformerManager.GetResourceLister(ftc.GetSourceTypeGVK(), clusterName)
+	if !exists {
+		return nil, nil, fmt.Errorf("failed to get dynamic client")
+	}
+
+	// If cluster cache is synced, we check the store.
+	// Otherwise, we will have to issue a list request.
+	if hasSynced() {
+		objects, err := resourceLister.List(labels.Everything())
+		if err != nil {
+			return nil, nil, err
+		}
+
+		clusterObjs = make([]*unstructured.Unstructured, len(objects))
+		for i := range objects {
+			clusterObjs[i] = objects[i].(*unstructured.Unstructured)
+			namespaces.Insert(clusterObjs[i].GetNamespace())
+		}
+	} else {
+		client, err := s.getClusterClient(clusterName)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		objects, err := client.Resource(ftc.GetSourceTypeGVR()).Namespace(corev1.NamespaceAll).List(ctx,
+			metav1.ListOptions{
+				LabelSelector: labels.SelectorFromSet(labels.Set{
+					managedlabel.ManagedByKubeAdmiralLabelKey: managedlabel.ManagedByKubeAdmiralLabelValue,
+				}).String(),
+			},
+		)
+		if err != nil {
+			// the CRD may be deleted before the CR, and then the member cluster will be stuck
+			// in the cascade deletion because the api no longer exists. So we need to ignore
+			// the NoMatch and NotFound error.
+			// Whether to return NoMatch or NotFound depends on whether the client has visited CR,
+			// if so, returns NotFound (because the client has a scheme cache), otherwise returns NoMatch.
+			if meta.IsNoMatchError(err) || apierrors.IsNotFound(err) {
+				return clusterObjs, namespaces, nil
+			}
+			return nil, nil, err
+		}
+
+		clusterObjs = make([]*unstructured.Unstructured, len(objects.Items))
+		for i := range objects.Items {
+			clusterObjs[i] = &objects.Items[i]
+			namespaces.Insert(clusterObjs[i].GetNamespace())
+		}
+	}
+	return clusterObjs, namespaces, nil
+}
+
+func (s *SyncController) deleteClusterObjects(
+	ctx context.Context,
+	ftc *fedcorev1a1.FederatedTypeConfig,
+	namespaces []string,
+	client dynamic.Interface,
+	clusterName string,
+) error {
+	sourceGVR := ftc.GetSourceTypeGVR()
+	deletionPropagation := metav1.DeletePropagationBackground
+	var methodNotSupportedErr error
+	if ftc.GetNamespaced() {
```

Comment:
Can we consider the namespaces of the cluster-scoped resources as `[]string{corev1.NamespaceAll}`, so that the two operations can be unified?

Reply:
The code has been updated.